### ISSUE TITLE
Improve non-cachable target because script's side effects

### ISSUE THREAD

**Issue by andresbrun** [tuist/tuist#6774]
### What problem or need do you have?

Even if I define input files and output files on the script phase like this:
<img width="866" alt="image" src="https://github.com/user-attachments/assets/1bcf36aa-a0d3-406f-b849-f7a6b5b9aa51">

I get the warning:
![image](https://github.com/user-attachments/assets/af62cc31-c571-45d1-ab35-f80637c8499e)

### Potential solution

If Tuist detects input and output files on the script it should trust the developer they are right and silent this warning.

### macOS version

14.6.1

### Tuist version

4.28.0

### Xcode version

16.0

**Comment by pepicrft** [comment]
@andresbrun note that we can't detect whether that script that you are referencing from, `Scripts/generate-core-images.sh`, has side effects. I'd assume it does: the images. So the warning seems accurate. What would you expect Tuist to do in such case, where we can't know about your side effects unless you are explicit about it?

**Comment by andresbrun** [comment]
I thougt I was being explicit by using input and output paths. The script uses sourcery to create the enum out of the xcassets. I understand Tuist cannot introspect and infers what the .sh does that's why I think it should trust the developer is doing the right thing and silent the warning in case input paths and output paths are defined. What other approach could we take?

**Comment by fortmarek** [comment]
Since there's no way to silence the warning, I think it's better to trust the developer here with the input and output paths, instead of showing a warning _every_ single time a project is generated.

Developers then learn to ignore warnings altogether, which is counterproductive.

**Comment by pepicrft** [comment]
> What other approach could we take?

> Since there's no way to silence the warning, I think it's better to trust the developer here with the input and output paths, instead of showing a warning every single time a project is generated.

I agree that we need an API. I think the API should educate the developer on a "side effect" means in this context so that caching works reliably for them and they don't disable it without reasoning about the implications of that decision:

```diff
.pre(
  script: "sh ....", 
+ systemDependencies: [] I explicitly say I don't depend on anything
)

.pre(
  script: "sh ....", 
+ systemDependencies: [
+  .envVariable("FOO")
+  .file("path/to/file")
+ ]
)
```

**Comment by fortmarek** [comment]
> I agree that we need an API. I think the API should educate the developer on a "side effect" means in this context so that caching works reliably for them and they don't disable it without reasoning about the implications of that decision:

Aren't we duplicating a bit the concept of input and output paths? If your script depends on stuff not declared in them, then your local builds might get stale in the same way as your binaries would.
